CMakeLists.txt

```diff
@@ -551,6 +552,37 @@ if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/nnvm/CMakeLists.txt")
   list(APPEND mxnet_LINKER_LIBS ${nnvm_LINKER_LIBS})
 endif()
 
+# ---[ NNPack
+if(USE_NNPACK)
+  if (USE_MKLDNN)
+    message(FATAL_ERROR "Either MKLDNN or NNPack can be enabled but not both.")
+  endif()
+  # Add in NNPack and its dependencies
+  set(NNPACK_SOURCE_DIR "${CMAKE_SOURCE_DIR}/3rdparty/NNPACK")
+  if (EXISTS "${NNPACK_SOURCE_DIR}")
+    if (GOOGLETEST_SOURCE_DIR AND EXISTS "${GOOGLETEST_SOURCE_DIR}")
+      set(GOOGLETEST_SOURCE_DIR "${GTEST_ROOT}" CACHE STRING "Google Test source directory")
+    endif()
+
+    # put NNPack dependencies in appropriate folders, see NNPack
+    # for other options
+    set(CONFU_DEPENDENCIES_SOURCE_DIR "${NNPACK_SOURCE_DIR}/deps"
+        CACHE PATH "Confu-style dependencies source directory")
+    set(CONFU_DEPENDENCIES_BINARY_DIR "${CMAKE_BINARY_DIR}/3rdparty/NNPACK/deps"
+        CACHE PATH "Confu-style dependencies source directory")
+
+    add_subdirectory("${NNPACK_SOURCE_DIR}")
```

Comment:
I'd suggest to do set the following variables:
```
set(NNPACK_BUILD_TESTS OFF CACHE BOOL "")
set(NNPACK_BUILD_BENCHMARKS OFF CACHE BOOL "")
set(NNPACK_LIBRARY_TYPE "static" CACHE STRING "")
set(PTHREADPOOL_LIBRARY_TYPE "static" CACHE STRING "")
set(CPUINFO_LIBRARY_TYPE "static" CACHE STRING "")
```
`NNPACK_BUILD_TESTS` and `NNPACK_BUILD_BENCHMARKS` disabled building NNPACK's own tests and benchmarks (which are enabled by default), and setting `NNPACK_LIBRARY_TYPE`, `PTHREADPOOL_LIBRARY_TYPE` and `CPUINFO_LIBRARY_TYPE` as static would build these projects as static libraries even if `BUILD_SHARED_LIBS` is `ON`.

After `add_subdirectory`, you'd likely want to add
```
set_property(TARGET nnpack PROPERTY POSITION_INDEPENDENT_CODE ON)
set_property(TARGET pthreadpool PROPERTY POSITION_INDEPENDENT_CODE ON)
set_property(TARGET cpuinfo PROPERTY POSITION_INDEPENDENT_CODE ON)
```
to build nnpack, pthreadpool, and cpuinfo with `-fPIC` even through they are static libraries.